Add cart & checkout block/shortcode info to tracker data

haszari opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR adds new data to the tracker snapshot. We want to understand the usage of the (upcoming) cart & checkout blocks, and their existing shortcode counterparts.

To help with this, this PR looks in the content for the cart & checkout pages for the shortcode or blocks, and adds yes/no values to the tracker data.

pb0Spc-ou-p2

### How to test the changes in this Pull Request:
Ensure your test site is opted into tracking (`WooCommerce > Settings > Advanced > WooCommerce.com > Usage Tracking`), and wait. Or, you can force an update using `wp shell`:

- `WC_Tracker::send_tracking_data( true );` 

You can also call the new method directly to check the results:

- `WC_Tracker::get_cart_checkout_info();` 

This should return an array with these keys:

```
=> array(4) {
  ["cart_page_contains_cart_block"]=>
  string(3) "Yes"
  ["cart_page_contains_cart_shortcode"]=>
  string(2) "No"
  ["checkout_page_contains_checkout_block"]=>
  string(3) "Yes"
  ["checkout_page_contains_checkout_shortcode"]=>
  string(2) "No"
}
```

Experiment with different pages set as cart/checkout pages (`WooCommerce > Settings > Advanced > Page setup`), and different content on those pages (e.g. cart or checkout shortcodes and/or blocks). 

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

haszari committed:
add cart info to tracker:
- adds a new key 'cart_checkout'
- adds boolean props indicating if store cart & checkout use blocks or shortcodes

haszari commented:
Not sure how to test this locally yet - will update the description with more info. If anyone knows a good way to locally test tracker snapshot, please comment :)

I've also tentatively assigned this to 4.1.0 milestone. Assuming we go with this approach, we'd want to include this in 4.1.0. Happy to discuss :)

deltaWhiskey commented on includes/class-wc-tracker.php:
```diff
@@ -621,6 +624,60 @@ private static function get_order_dates() {
+		$result = $wpdb->get_var(
+			$wpdb->prepare(
+				"
+				SELECT COUNT( * ) FROM {$wpdb->prefix}_posts
```
I believe this should be `{$wpdb->prefix}posts`, like line 593. When I run this method in `wpcli` I get an error message about table `wp__posts` not existing.

haszari replied:
Now fixed in 0b79a65829d03ccf8a9d09774be685862c2bf3a5, thanks for spotting :)

deltaWhiskey reviewed (changes requested):
One way to test this is `wp eval 'print_r(WC_Tracker::get_cart_checkout_info(), true);'`. The same thing can be done with `get_tracking_data()` if you want to see how it sits with the rest of the data.

The general approach looks good, but there's an issue with some SQL.

haszari committed:
fix various issues with post_contains_text:
- correct table name / prefix
- ensure search string has wildcards before & after
- return value is '0' or '1', fix return logic (was returning 'Yes' no
matter what)

haszari commented:
@deltaWhiskey I've fixed the SQL issue and a few other problems in that function, it's now tested and working well for me. Please re-review when you have time :)

haszari commented:
@vedanshujain @peterfabian Can I get a proton review of this please – keen to make sure it's shipshape in time for the 4.1 code freeze. Let me know there's anything I need to do or tidy up :)

vedanshujain commented on includes/class-wc-tracker.php:
```diff
@@ -621,6 +624,63 @@ private static function get_order_dates() {
+		global $wpdb;
+
+		$cart_page_id     = wc_get_page_id( 'cart' );
+		$checkout_page_id = wc_get_page_id( 'checkout' );
```
Note that this could return -1, which means cart / checkout page is not properly setup. Should we account for this as a different return value to not mess up the stats?

vedanshujain commented on includes/class-wc-tracker.php:
```diff
@@ -621,6 +624,63 @@ private static function get_order_dates() {
+
+		$cart_page_id     = wc_get_page_id( 'cart' );
+		$checkout_page_id = wc_get_page_id( 'checkout' );
+		return array(
```
note that we can save couple of queries here by doing the matching in php instead of SQL, but since this function runs in cron, it doesn't matter much.

vedanshujain commented on includes/class-wc-tracker.php:
```diff
@@ -621,6 +624,63 @@ private static function get_order_dates() {
+		global $wpdb;
+
+		$cart_page_id     = wc_get_page_id( 'cart' );
+		$checkout_page_id = wc_get_page_id( 'checkout' );
```
Note that this could return -1, which means cart / checkout page is not properly setup. Should we account for this as a different return value to not mess up the stats?

haszari replied:
That's a nice idea – perhaps we can add that as another data point later (if useful). I haven't closely reviewed the other data in tracker already, we might track whether these pages are set up or onboarding is completed already.

vedanshujain reviewed (approved):
LGTM! Left a question (feel free to ignore the query comment - not a blocker), other than that looks good.

haszari commented:
@vedanshujain I don't have permissions to merge - can you or @woocommerce/proton merge on my behalf please? Thank you :)
